### ISSUE TITLE
IPP printing

### ISSUE THREAD

**Issue by nerijus** [BenoitDuffez/AndroidCupsPrint#40]
On my Linux PC cups uses Device URI: ipp://192.168.1.22/ipp
If I enter the same URI in Android CUPS Print, the printer is added, but when I try to print it says "Printer is not responding".

**Comment by nerijus** [comment]
If I use URI http://192.168.1.22:631/ipp, I get "The printer name was not found on the server (HTTP 404). please check its name."

**Comment by nerijus** [comment]
In logcat I see:
Caught exception while connecting to printer http://192.168.1.22:631/printers/: http://192.168.1.22:631/printers/
HTTP response code: 404

I can open http://192.168.1.22:631/ipp in a browser (it opens blank page), but http://192.168.1.22:631/printers shows "404 Not Found". My printer is connected via old HP JetDirect device, and it seems it does not support "/printers" in URL. Is it possible to make it work?

**Comment by BenoitDuffez** [comment]
I'm not sure the app works with plain IPP (anymore?). You need to have a CUPS server to get this app working. That's what I use for development, so if you differ from my setup, it's 100% untested :(

**Comment by nerijus** [comment]
I was able to print successfully using CUPS server.
Is it possible to make it work with plain IPP? Android CUPS Print could not search for printers if I enter printer URL and Name manually.

**Comment by BenoitDuffez** [comment]
I'm glad it works with CUPS. I have only tested this app with a CUPS server (with/without SSL and with/without Basic Auth). Other than that, it's just legacy code from the original writer. I don't have time right now to work on this.
It would be possible to make it work (again?) over IPP but I won't do it, at least not right now.
As per the manual input, there are two fields: [one for the IP/host + search button](https://github.com/BenoitDuffez/AndroidCupsPrint/blob/master/app/src/main/java/io/github/benoitduffez/cupsprint/AddPrintersActivity.java#L92), that will scan; [the other is static add by URL](https://github.com/BenoitDuffez/AndroidCupsPrint/blob/master/app/src/main/java/io/github/benoitduffez/cupsprint/AddPrintersActivity.java#L60).

**Comment by nerijus** [comment]
Yes, but adding static URL still does searching too, it still wants to find printers at http://192.168.1.22:631/printers/, find default printer, etc. I tried to patch source not to do that, but no success yet.

**Comment by BenoitDuffez** [comment]
Adding a static URL saves it to the preferences files, which is loaded when Android does the discovery.
So if your CUPS server replies to mDNS queries, there will be both: discovery result + manual entry.
That's per design.
If you don't want that, either disable mDNS on the CUPS server, or disable the mDNS scans in the app's source.

**Comment by nerijus** [comment]
I didn't have CUPS server enabled before, but it still searched for printers at the .../printers/ URL. How do I disable mDNS scans in the app's source?

**Comment by BenoitDuffez** [comment]
It's in [`CupsPrinterDiscoverySession#scanPrinters()`](https://github.com/BenoitDuffez/AndroidCupsPrint/blob/master/app/src/main/java/io/github/benoitduffez/cupsprint/printservice/CupsPrinterDiscoverySession.java#L300).

**Comment by BenoitDuffez** [comment]
I guess the changes I made allow for custom path (e.g. `/ipp` instead of `/printers/Printer_Name`).

If you can build, can you tell me if this fixes your issue? Otherwise you'll have to wait for the next beta upload on the play store.

**Comment by nerijus** [comment]
I built the develop branch. It seems it fixes the issue so that it does not search for printers at the .../printers/ URL for manually added printer.
When trying to print I get "Printer is not responding", but thinking more about it, I will need a client side printer driver, otherwise I can print text only, because HP JetDirect is not a full cups server, so it's probably not worth investigating.

Just for the testing - I was able to print successfully using CUPS server, as before.
